internals/overlord/logstate/loki.go

```diff
@@ -0,0 +1,148 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package logstate
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/canonical/pebble/internals/plan"
+	"github.com/canonical/pebble/internals/servicelog"
+)
+
+type lokiClient struct {
+	target    *plan.LogTarget
+	remoteURL string
+	// buffered entries are "sharded" by service name
+	entries map[string][]lokiEntry
+}
+
+func newLokiClient(target *plan.LogTarget) logClient {
+	return &lokiClient{
+		target:    target,
+		remoteURL: target.Location,
+		entries:   map[string][]lokiEntry{},
+	}
+}
+
+func (c *lokiClient) Write(ctx context.Context, entry servicelog.Entry) error {
+	c.entries[entry.Service] = append(c.entries[entry.Service], asLokiEntry(entry))
+	if c.numEntries() >= 10 {
+		return c.Flush(ctx)
+	}
+	return nil
+}
+
+func (c *lokiClient) numEntries() int {
+	sum := 0
+	for _, l := range c.entries {
+		sum += len(l)
+	}
+	return sum
+}
+
+func asLokiEntry(entry servicelog.Entry) lokiEntry {
+	return lokiEntry{
+		strconv.FormatInt(entry.Time.UnixNano(), 10),
+		strings.TrimRight(entry.Message, "\n"),
+	}
+}
+
+func (c *lokiClient) Flush(ctx context.Context) error {
+	if c.numEntries() == 0 {
+		return nil // no-op
+	}
+	defer c.emptyBuffer()
+
+	req := c.buildRequest()
+	jsonReq, err := json.Marshal(req)
+	if err != nil {
+		return fmt.Errorf("encoding request to JSON: %v", err)
+	}
+
+	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, c.remoteURL, bytes.NewReader(jsonReq))
+	if err != nil {
+		return fmt.Errorf("creating HTTP request: %v", err)
+	}
+	httpReq.Header.Set("Content-Type", "application/json")
+
+	resp, err := http.DefaultClient.Do(httpReq)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = resp.Body.Close() }()
+
+	// Check response status code to see if it was successful
+	if sc := resp.StatusCode; sc < 200 || sc >= 300 {
```

Comment:
as a first pass, let's just drop the request if the return code is not 2xx. We can think about retry logic in future iterations.

The possible issue with retrying is logs getting backed up - while we're still trying to send old logs, we may be forced to drop new logs if it takes too long.

Reply:
We should handle a 429 properly, so we don't continue spamming the server when it has asked us not to.

Two ideas:
1. In subsequent `Flush` calls, the client should block until the `Retry-After` timeout has elapsed.
2. Pass the `Retry-After` value up to the gatherer, so it can adjust its timer accordingly.

Probably (2) is the "right" way to do this, but it may add a bit of complexity, so (1) might be an acceptable compromise.

Reply:
@benhoyt I added backoff logic throughout the client and gatherer